util/src/main/java/org/killbill/billing/util/discovery/provider/KillbillEurekaInstanceInfoProvider.java

```diff
@@ -0,0 +1,128 @@
+                                                                  .setRenewalIntervalInSecs(config.getLeaseRenewalIntervalInSeconds())
+                                                                  .setDurationInSecs(config.getLeaseExpirationDurationInSeconds());
+
+            if (vipAddressResolver == null) {
+                vipAddressResolver = new Archaius1VipAddressResolver();
+            }
+
+            // Builder the instance information to be registered with eureka server
+            InstanceInfo.Builder builder = InstanceInfo.Builder.newBuilder(vipAddressResolver);
+
+            // set the appropriate id for the InstanceInfo, falling back to datacenter Id if applicable, else hostname
+            String instanceId = config.getInstanceId();
+            DataCenterInfo dataCenterInfo = config.getDataCenterInfo();
+            if (instanceId == null || instanceId.isEmpty()) {
+                if (dataCenterInfo instanceof UniqueIdentifier) {
+                    instanceId = ((UniqueIdentifier) dataCenterInfo).getId();
+                } else {
+                    instanceId = config.getHostName(false);
+                }
+            }
+
+            String defaultAddress = config.getIpAddress();
+
+            builder.setNamespace(config.getNamespace())
+                   .setInstanceId(instanceId)
+                   .setAppName(config.getAppname())
+                   .setAppGroupName(config.getAppGroupName())
+                   .setDataCenterInfo(config.getDataCenterInfo())
+                   .setIPAddr(config.getIpAddress())
+                   .setHostName(defaultAddress)
+                   .setPort(config.getNonSecurePort())
+                   .enablePort(PortType.UNSECURE, config.isNonSecurePortEnabled())
+                   .setSecurePort(config.getSecurePort())
+                   .enablePort(PortType.SECURE, config.getSecurePortEnabled())
+                   .setVIPAddress(config.getVirtualHostName())
+                   .setSecureVIPAddress(config.getSecureVirtualHostName())
+                   .setHomePageUrl(config.getHomePageUrlPath(), config.getHomePageUrl())
+                   .setStatusPageUrl(config.getStatusPageUrlPath(), config.getStatusPageUrl())
+                   .setASGName(config.getASGName())
+                   .setHealthCheckUrls(config.getHealthCheckUrlPath(),
+                                       config.getHealthCheckUrl(), config.getSecureHealthCheckUrl());
+
+
+            // Start off with the STARTING state to avoid traffic
+            if (!config.isInstanceEnabledOnit()) {
+                InstanceStatus initialStatus = InstanceStatus.STARTING;
+                LOG.info("Setting initial instance status as: " + initialStatus);
+                builder.setStatus(initialStatus);
+            } else {
+                LOG.info("Setting initial instance status as: {}. This may be too early for the instance to advertise "
```

Comment:
I am indeed wondering if this shouldn't be more tightly integrated with our healthcheck. For instance, when calling `KillbillHealthcheck.putOutOfRotation`, you would probably send `OUT_OF_SERVICE` to Eureka?

One way to do this could be to have an optionally injected field in `KillbillHealthcheck`: `List<ServiceDiscovery>` where `ServiceDiscovery` is a generic interface which exposes a few methods for broadcasting (we can get inspired by `ApplicationInfoManager`). When injected, the healthcheck would also call the service discoveries methods (one implementation would be the Eureka one) and you wouldn't even need to worry about initialization/shutdown sequence (already done for you, and even controllable by JMX).

If we go down that road, we could also move the Eureka-specific code to a new Maven module in `killbill-platform`, which would solve some dependencies issues mentioned above (Kill Bill wouldn't have any dependency on Eureka, unless you add that jar in your webapp). That being said, in your environment, the dependencies conflict do exist and you have no idea at runtime which jar is being used 😬Best would probably be to have that new module produce a shaded uber jar.

(/cc @sbrossie)